¡Hola José!

Mi nombre es Facundo Lozano! Un gusto conocerte, seré tu revisor en este proyecto.

Cuando vea un error la primera vez, simplemente lo señalaré. Te dejaré encontrarlo y arreglarlo tú mismo. Además, a lo largo del texto haré algunas observaciones sobre mejora en el código y también haré comentarios sobre tus percepciones sobre el tema. Pero si aún no puedes realizar esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a comentarios y discusiones sobre el tema.

Encontrará mis comentarios a continuación: **no los mueva, modifique ni elimine**.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como este:

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Exito. Todo se ha hecho de forma exitosa.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Observación. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos. 
</div>

Puedes responder utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>
</div>

# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 1) </b> <a class="tocSkip"></a>

José, quería tomarme este tiempo al inicio de tu proyecto para comentarte mis apreciaciones generales de esta iteración de tu entrega.

Antes que nada quería darte la bienvenida al mundo de los datos, te deseo lo mejor y espero que consigas lograr tus objetivos. Personalmente me gustaría brindarte el siguiente consejo, "Está bien equivocarse, es normal y es lo mejor que te puede pasar. Aprendemos de los errores y eso te hará mejor programadora ya que podrás descubrir cosas a medida que avances y son estas cosas las que te darán esa experiencia para ser una gran cientifica de datos."
    
Ahora si yendo a esta notebook. Quiero felicitarte porque he visto que tienes un gran manejo de python y de lógica a la hora de resolver ciertos problemas, sin embargo he encontrado algunos errores que estoy seguro que podrás resolverlos. En general has estado muy bien así que a no preocuparse que en la siguiente iteración estoy seguro que se aprueba.
</div>

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv('/datasets/music_project_en.csv')

Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`Respetar el valor estilo snake_case, en este caso en la primera columna, quedaría como user_id.




<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Bien hecho José! Haz realizado todos los pasos de manera excelente, importaste, cargaste, asignaste y mostraste los datos tal como se debe hacer.</div>

### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

        R: Los tipos de datos que tenemos son de tipo string, y fecha. Podemos entender lo que almacenan gracias al nombre la columna es por eso que es muy importante que el nombre la columna esté relacionada con la información que contiene.

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

        R: No, porque los pasos anteriores solo nos proporciona una vista general de los datos. Por lo tanto, requerimos más información.

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

        R: Además del estilo de los encabezados, gracias al método info() podemos saber cuantos datos ausentes hay, por ejemplo, en Track, artist y genre. Aún no podemos observar datos duplicados o datos incorrectos porque aún no aplicamos preprocesameinto de datos.

<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente José, en general tenes una comprensión increíble, sin embargo detecté un pequeño error en tu primera respuesta. Si bien, en gran parte podemos deducir los tipos de datos que almacenan según el nombre de la columna, esto no es siempre así. Por ejemplo: Columna "Edad" y el tipo de datos son efectivamente números pero de tipo string "22". </div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy bien! Has entendido que no es necesario utilizar el método print().</div>

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle en los encabezados poniendo todo en minúsculas
new_names_lowered = []
for old_name in df.columns:
    name_lowered = old_name.lower()
    new_names_lowered.append(name_lowered)
new_names_lowered

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']

<div class="alert alert-block alert-danger">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

A no preocuparse! Vienes muy bien pero en la celda anterior estamos obteniendo una lista con los nombres de las columnas y lo que debemos hacer es aplicar estos cambios en el dataframe, para ello podrías pensar en un método nativo de pandas como .rename(), investigalo!
</div>

Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [7]:
# Bucle en los encabezados eliminando los espacios
new_names_no_spaces = []
for old_names in new_names_lowered:
    name_stripped = old_names.strip()
    new_names_no_spaces.append(name_stripped)
new_names_no_spaces

['userid', 'track', 'artist', 'genre', 'city', 'time', 'day']

<div class="alert alert-block alert-danger">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Si arreglamos lo anterior esto es igual! Venimos excelente y de los errores se aprende, aprovecha el bucle que te brinda el nombre de la columna para en cada iteración e intenta aplicar el método que te mencionaba anteriormente. Y recuerda que estos cambios se deben realizar directamente sobre el dataframe.
</div>

Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambiar el nombre de la columna "userid"
new_names = []
for old_names in df.columns:
    name_stripped = old_names.strip()
    name_lowered = name_stripped.lower()
    name_no_spaces = name_lowered.replace(' ','_')
    new_names.append(name_no_spaces)
print(new_names)

new_names[0] = new_names[0].replace('userid','user_id')

print(new_names)

df.columns = new_names

['userid', 'track', 'artist', 'genre', 'city', 'time', 'day']
['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']


Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprobar el resultado: la lista de encabezados
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente José, en este caso si aplicaste dichos cambios sobre el dataframe. Comprendo que tal vez la interpretación fuese aplicar estos cambios al final pero estos ejercicios demandan aplicarlos sobre el momento por lo que te invito a realizar esos pequeños cambios. Por último felicitaciones por el método que utilizaste para cambiar las columnas aunque esto nos obligará siempre a necesitar una lista con el mismo length que nuestras columnas, sigue investigando los beneficios de .rename()</div>

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [10]:
# Calcular el número de valores ausentes
df.isnull().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Felicitaciones José, aplicaste un doble método para conseguir tu resultado. Si bien esta perfecto porque responde la consigna, te recomiendo investigar el método .isna() </div>

No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [47]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track','artist','genre']
for col in columns_to_replace:
    df[col].fillna('unknown',inplace=True)

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Felicitaciones! Implementaste una lógica de gran utilidad para evitar repeteciones en tu código, perfecto José!.</div>

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [48]:
# Contar valores ausentes
df.isnull().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Bien hecho, al igual que te comenté anteriormente aplicaste un doble método para conseguir tu resultado. Responde la consigna y es lo que importa, puedes investigar el método .isna() si deseas. </div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [52]:
# Contar duplicados explícitos
df.duplicated().sum()

3826

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Bien resuelto! Simple y conciso, perfecta la utilización del doble método corroborando los duplicados y sumandolos con sum(). </div>

Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [54]:
# Eliminar duplicados explícitos
df.drop_duplicates(inplace=True)

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy bien, Data Scientist nato! Excelente implementación de drop_diplicates(), sigue así José!.</div>

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [55]:
# Comprobar de nuevo si hay duplicados
df.duplicated().sum()


0

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Seguimos en racha! Veo que ya manejas a la perfección la implementación de duplicated() y sum(). A seguir por más! </div>

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [56]:
# Inspeccionar los nombres de géneros únicos
df['genre'].unique()

array(['rock', 'pop', 'folk', 'dance', 'rusrap', 'ruspop', 'world',
       'electronic', 'unknown', 'alternative', 'children', 'rnb', 'hip',
       'jazz', 'postrock', 'latin', 'classical', 'metal', 'reggae',
       'triphop', 'blues', 'instrumental', 'rusrock', 'dnb', 'türk',
       'post', 'country', 'psychedelic', 'conjazz', 'indie',
       'posthardcore', 'local', 'avantgarde', 'punk', 'videogame',
       'techno', 'house', 'christmas', 'melodic', 'caucasian',
       'reggaeton', 'soundtrack', 'singer', 'ska', 'salsa', 'ambient',
       'film', 'western', 'rap', 'beats', "hard'n'heavy", 'progmetal',
       'minimal', 'tropical', 'contemporary', 'new', 'soul', 'holiday',
       'german', 'jpop', 'spiritual', 'urban', 'gospel', 'nujazz',
       'folkmetal', 'trance', 'miscellaneous', 'anime', 'hardcore',
       'progressive', 'korean', 'numetal', 'vocal', 'estrada', 'tango',
       'loungeelectronic', 'classicmetal', 'dubstep', 'club', 'deep',
       'southern', 'black', 'folkrock', 

<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Buena implemetación! Bien hecho, a este paso le podríamos agregar el método .sorted() que nos sería de gran utilidad ya que nos permitiría ver con un orden alfabético dichas columnas. Imagina que hubo un error y tenemos dos valores "garage" y "garagee", son distintos aunque nosotros sabemos que son lo mismo. Investiga .sorted()</div>

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [14]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'].replace(wrong_genre,correct_genre,inplace=True)
    return df

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [15]:
# Eliminar duplicados implícitos
df = replace_wrong_genres(['hip','hop','hiphop'], 'hip-hop')

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [16]:
# Comprobación de duplicados implícitos
df['genre'].unique()

array(['rock', 'pop', 'folk', 'dance', 'rusrap', 'ruspop', 'world',
       'electronic', nan, 'alternative', 'children', 'rnb', 'hip-hop',
       'jazz', 'postrock', 'latin', 'classical', 'metal', 'reggae',
       'triphop', 'blues', 'instrumental', 'rusrock', 'dnb', 'türk',
       'post', 'country', 'psychedelic', 'conjazz', 'indie',
       'posthardcore', 'local', 'avantgarde', 'punk', 'videogame',
       'techno', 'house', 'christmas', 'melodic', 'caucasian',
       'reggaeton', 'soundtrack', 'singer', 'ska', 'salsa', 'ambient',
       'film', 'western', 'rap', 'beats', "hard'n'heavy", 'progmetal',
       'minimal', 'tropical', 'contemporary', 'new', 'soul', 'holiday',
       'german', 'jpop', 'spiritual', 'urban', 'gospel', 'nujazz',
       'folkmetal', 'trance', 'miscellaneous', 'anime', 'hardcore',
       'progressive', 'korean', 'numetal', 'vocal', 'estrada', 'tango',
       'loungeelectronic', 'classicmetal', 'dubstep', 'club', 'deep',
       'southern', 'black', 'folkrock', 'f

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Felicitaciones José! Aplicaste todo correctamente y tu función cumple con los requisitos. Además me gustaría agregar tu prolijidad a la hora de expresar tu código. Bien hecho! </div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`


Analizar duplicados explicítos es relativamente sencillo. Ejecutando los métodos dulpicated(), sum(), drop_duplicates(), habremos sabido cuantos duplicados hay y podemos quitar en el dataset.

Analizar duplicados implícitos es una tarea más complicada. Debido a que puede referise a la misma palabra o etiqueta pero que esté mal escrita y pueden ser miles de datos casi imposibles de identificar a simple vista.


No necesariamente que un dato esté vacío signifique que sea inútil y con una etiqueta lo podemos sustituir. De preferencia que sepamos identificar fácilmente como 0 o 'desconocido'.

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente! A medida que continúes con tus estudios aprenderás técnicas para como manejar estos casos de valores nulos pero que aún así pueden ser de utilidad según sea el manejo de los mismos. Nunca debemos creer que un valor nulo conlleva a una eliminación directa, diferente es hablar de observaciones duplicadas en su totalidad. </div>

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [79]:
# Contar las canciones reproducidas en cada ciudad
df.groupby(by='city')['track'].count()

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Felicitaciones José! Una implementación perfecta del .groupby(), este es un método de mucha utilidad que se aplica en todo proyecto de Data Science así que es importante tener un buen manejo sobre este. A seguir que vienes genial! </div>

`Comenta tus observaciones aquí`

Aplicando el método dividir-aplicar-combinar fue útil al saber cuántas canciones se consumen por ciudad.

En Shelbyville se reprodujeron 18512 y en Springfield 42741.

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [81]:
# Calcular las canciones reproducidas en cada uno de los tres días
df.groupby(by='day')['track'].count()

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Misma implementación, diferente agrupación. Bien hecho! </div>

`Comenta tus observaciones aquí`

El día viernes se consumió 21 840 canciones, en lunes 21 354 y el miercoles 18 059.

In [87]:
df[df['day'] == 'Friday']

user_id                        track                artist  \
1      55204538  Delayed Because of Accident      Andreas Rönnberg   
5      842029A1                       Chains              Obladaet   
8      8FA1D3BE                     L’estate           Julia Dalia   
13     CC782B0F         After School Special  Detroit Grand Pubahs   
30     EBA5A8B2       Päivän Nousu Nostajani             Värttinä   
...         ...                          ...                   ...   
65071  92378E24    Swing it Like You Mean it                OJOJOJ   
65072  C532021D       We Can Not Be Silenced                Pänzer   
65076  C5E3A0D5                    Jalopiina               unknown   
65077  321D0506                Freight Train         Chas McDevitt   
65078  3A64EF84    Tell Me Sweet Little Lies          Monica Lopez   

              genre         city      time     day  
1              rock  Springfield  14:07:09  Friday  
5            rusrap  Shelbyville  13:09:41  Friday  
8            ruspop  Springfield  09:17:40  Friday  
13            dance  Springfield  20:04:12  Friday  
30            world  Shelbyville  08:00:31  Friday  
...             ...          ...       ...     ...  
65071        techno  Springfield  21:12:56  Friday  
65072  extrememetal  Springfield  08:38:24  Friday  
65076    industrial  Springfield  20:09:26  Friday  
65077          rock  Springfield  21:43:59  Friday  
65078       country  Springfield  21:59:46  Friday  

[21840 rows x 7 columns]

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [91]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    df_day = df[df['day'] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    df_city = df_day[df_day['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    df_id = df_city['user_id'].count()
    # Devolve el número de valores de la columna 'user_id'
    return df_id

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente forma de resolverlo! Realizaste 2 filtros sequenciales que te brindarían el resultado posteriormente al utilizar .count(), bien hecho!. ¿Sabías que podes establecer 2 filtros en un mismo llamado? De forma opcional si quieres puedes investigar como hacer esto.</div>

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [92]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks('Monday','Springfield')

15740

In [93]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks('Monday','Shelbyville')

5614

In [94]:
# El número de canciones reproducidas en Springfield el miércoles
number_tracks('Wednesday','Springfield')

11056

In [95]:
# El número de canciones reproducidas en Shelbyville el miércoles
number_tracks('Wednesday','Shelbyville')

7003

In [96]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks('Friday','Springfield')

15945

In [97]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks('Friday', 'Shelbyville')

5895

**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`

Rechazaría la hipótesis. Aunque, En Springfield se consume más música que en Shelbyville. esto no me indica totalmente que existen diferencias en que forma que consumen música. 

Faltaría analizar de qué género y en que momento se escucha una canción por ciudad y por día. Esto nos ayudará a saber si una persona prefiere escuchar rock mientras come y trabaja o pop antes de dormir. Esto si ayudaría a saber la deferencias en la forma en que los usuarios consumen música.

También un dato importante sería conocer la densidad poblacional de Sprinfield y Shelbyville para tener una conclusión más completa.

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`

Primero, las hipótesis son una gran inicio de partida al momento de analizar datos para obtener preguntas y contestarlas. 
Posterioremente, analizar la hipóstesis para saber si la preguntas que se elaboraron y respondieron son las correctas para la misma.
Una vez hecho esto; sobre la misma hipótesis, crear nuevas preguntas.
Finalmente; obtener conclusiones que justifiquen la aprobación o no de la hipótesis.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)

<div class="alert alert-block alert-success">
<b>Comentario final. (Iteración 1) </b> <a class="tocSkip"></a>

Felicitaciones José, hemos llegado al final de la notebook y tu desempeño fue increíble. Estas conclusiones reflejan y demuestran tu compromiso con el trabajo y lo valoramos. Si bien el proyecto se rechaza por 2 errores rojos que probablemente se deban a un error de interpretación, estoy seguro que en nuestra proxima iteración ya podrás seguir adelante. Espero con ansias tus correciones.
</div>